Support for *flat mapped* coherent secure RAM

etienne-lms opened the pull request:
This change enables secure coherent memory support and allow platform vexpress-qemu_virt to easily define a coherent memory at top of the secure RAM.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -53,9 +53,27 @@
+ * Start coherent RAM with a platform specific structure.
+ * Generic data to be place in coherent RAM can use attribute "coherent_ram".
+ */
+#define DECLARE_COHERENT_RAM_SECTION						\
```
Why do we need a macro for this?

etienne-lms replied:
To add some flexibility regarding the location of the coherent RAM, in case one wants it at a dedicated place. I proposed to possible location: before the load address or after the core reserved vaddr range (virt, no phys only, to fit with pager constraints). Since there were those 2 places, I preferred to use a single macro. You feel it bring complexity ?

jenswi-linaro replied:
Aha, I didn't notice that. This is good then.
The `\` at the end of each line seems to have ended up at column 81 which obviously is beyond 80.

etienne-lms replied:
Yep, i've noticed that from travis feedback. Bad luck.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -356,3 +378,8 @@ SECTIONS
+
+#if defined(CFG_TEE_COHERENT_START)
+ASSERT((__coherent_start & (4096 - 1)) == 0, "Coherent RAM start is not 4Kbyte aligned")
+ASSERT((__coherent_end & (4096 - 1)) == 0, "Coherent RAM end is not 4Kbyte aligned")
```
Testing that `. = ALIGN(4096);` works?

etienne-lms replied:
This is extra checks.  Actually `__coherent_end` is already aligned, and static mapping will assert `CFG_TEE_COHERENT_START` (hence `__coherent_start`) is already 4kB page aligned.
Those checks aimed at detecting issues at build time.

etienne-lms replied:
Note: i will remove those extra checks.

jenswi-linaro replied:
The assert that `. = ALIGN(4096);` works is just silly and and other assert could be done inside the `DECLARE_COHERENT_RAM_SECTION` macro instead.

etienne-lms replied:
Adding several ASSERT() inside the macro `DECLARE_COHERENT_RAM_SECTION` combine with the 80 chars max line size make the macro not very easy to read. I will move all assertion at the end of the file. Nevertheless, fell free to comment back...

jenswi-linaro commented on core/arch/arm/plat-vexpress/platform_config.h:
```diff
@@ -222,6 +222,18 @@
+ * the secure RAM.
+ */
+#ifndef CFG_TEE_COHERENT_SIZE
+#define CFG_TEE_COHERENT_SIZE	0
```
Why is this needed?

etienne-lms replied:
Default: no coherent RAM => null sized.

jenswi-linaro replied:
May I suggest `CFG_TEE_COHERENT_SIZE ?= 0` in `core/arch/arm/arm.mk` instead?

etienne-lms replied:
Good idea. Thanks.

jenswi-linaro commented on lib/libutils/ext/include/compiler.h:
```diff
@@ -54,6 +54,7 @@
 #define __rodata_unpaged __section(".rodata.__unpaged")
 #define __early_bss	__section(".early_bss")
 #define __noprof	__attribute__((no_instrument_function))
+#define __coherent	__attribute__((coherent_ram))
```
My google karma is failing me, do you have a link to that attribute?

etienne-lms replied:
hmmm, will fix to `__section(".coherent_ram")`. Thanks.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -53,9 +53,27 @@
+		. = CFG_TEE_COHERENT_START;					\
+		.coherent (NOLOAD) : {						\
+			__coherent_start = .;					\
+			KEEP(*(coherent_ram));					\
```
All the other sections starts with a "." can't we have the same here?

etienne-lms replied:
my fault. i'll fix (see also comments on attribute `__coherent_ram`).

etienne-lms commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -53,9 +53,27 @@
 OUTPUT_ARCH(CFG_KERN_LINKER_ARCH)
 
+/*
+ * Start coherent RAM with a platform specific structure.
```
Oups, this comment is related to the libpsci integ. I will remove it.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -379,7 +374,11 @@ SECTIONS
+#if CFG_TEE_COHERENT_SIZE
+ASSERT(!(__coherent_start & (4096 - 1)),
+					"Coherent memory start alignment");
+ASSERT(!(CFG_TEE_COHERENT_SIZE & (4096 - 1)),
```
This and the following assert are useless, they are just checking that ` . = ALIGN(4096)` works.

etienne-lms replied:
The test above explicitly notifies developer that the coherent size follows the mapping constraint. Otherwise, this will be detected only a runtime. I found this explicit trace more convenient, but I agree it is redundant with the runtime assertions/panics.

Test below does not check the alignment but the size. In case one defines CFG_TEE_COHERENT_SIZE to a value below the effective coherent section computed at link stage, the coherent might the only partially mapped and only run time data abort could detect it.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -379,7 +374,11 @@ SECTIONS
-ASSERT((__coherent_start & (4096 - 1)) == 0, "Coherent RAM start is not 4Kbyte aligned")
-ASSERT((__coherent_end & (4096 - 1)) == 0, "Coherent RAM end is not 4Kbyte aligned")
+#if CFG_TEE_COHERENT_SIZE
+ASSERT(!(__coherent_start & (4096 - 1)),
```
I don't see the problem with wrapping this inside the `DECLARE_COHERENT_RAM_SECTION` macros as:
```
                ASSERT(!(__coherent_start & (4096 - 1)),    \
                       "Coherent memory start alignment");     \
```

etienne-lms replied:
An assertion in the macro could be ok, but the 3 make the macro a bit ugly (in my opinion).
But ok, I will move them (at least those really required) inside the macro.

etienne-lms commented on core/arch/arm/plat-vexpress/platform_config.h:
```diff
@@ -222,7 +222,6 @@
 #define GICD_OFFSET		0
 #define GICC_OFFSET		0x10000
 
-
```
minor: useless change, to be discarded.

etienne-lms commented:
Since the overall changes are quite small, I will rebase and squash the proposed commit to get a (hopefully) nice travis status and ease later reviews.

etienne-lms commented:
thanks travis! i missed generic cases: must fix platform not defining any `CFG_TEE_COHERENT_START`.
I also found I forgot to push a change and let few crappy stuff on default location of the coherent memory. Fixes on-going...

etienne-lms commented:
I will put this change on hold. 
My initial implementation was based on https://github.com/OP-TEE/optee_os/pull/1459 and I see now that it would imply too many conflicting changes to try to push it aside from this reference PR. Hence, I'll wait https://github.com/OP-TEE/optee_os/pull/1459 to mature and be merged before pushing back support for coherent memory.

etienne-lms commented:
Rebased and adapted (on top of recently merged https://github.com/OP-TEE/optee_os/pull/1459).
Updated content: coherent memory is now always located before the core load address.

etienne-lms commented:
rebased.

jforissier commented:
Looks like this needs some rework?

etienne-lms commented:
Yes, this deserves a rebase. Latest optee_os already defines coherent memory to maps rwx ram.
This p-r aimed at defining coherent mem (at least rw) and providing generic linking and mapping of the  coherent memory. Current p-r location coherent ram inside the tee_ram fla map constriant, for simplicity. Obviously https://github.com/OP-TEE/optee_os/pull/1729 locates coherent ram outside.

What's needs to be rebased:
- linking: declare a single coherent ram in **kern.lds.S** which can be above or below the flat mapped tee_ram) and define a `.coherent_ram` section attribute to allow linking code with the section.
- mapping: `register_phys_mem(COHERENT_MEM, ...);` should get the coherent mem to be mapped. This implies changes in [init_mem_map()#L845](https://github.com/OP-TEE/optee_os/blob/59ac3927b3bd14e133f57e7aa59278b09c2ca95d/core/arch/arm/mm/core_mmu.c#L845) to build a core vmem layout around coherent mem, `CFG_TEE_RAM_VA_SIZE` mapped and userland constraints.

etienne-lms commented:
Rebased. 
- `init_mem_map()` sequence that defines the virtual memory layout is modified.
- `register_phys_mem(TEE_MEM_COHERENT)` maps the coherent.
- Defining `CFG_TEE_COHERENT_START/_SIZE` gets a `__coherent` section linked.
  (used by libpsci)

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -842,49 +900,118 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+	 * To speed up scans of the memory_map table, it is sorted, moving
+	 * all small-page mapped area before the pgdir mapped areas.
+	 */
+	qsort(memory_map, last, sizeof(struct tee_mmap_region),
```
I see your point. I will remove this sorting.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -842,49 +900,118 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+
+#ifdef CFG_WITH_LPAE
+	/* LPAE: 1 pgdir for protection, user memory located above the core */
+	va = CORE_MMU_PGDIR_SIZE;
```
> So we start to map the rest at 2 MiB regardless of where flat mapped range is (it could even conflict with this)? 

The loop right after this init skips the virtual range already used to assigned virtual locations. This virtual area is expected to be fully cover the current value of `vaspace_start`/`vaspace_size`.

> More efficient use of level 2 tables would be to keep in same level 2 table as used above, it's 1 GiB so there should be plenty of space.

Agree. I will use the full 1GB around the flat map area to reuse the level2 xlat table.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -90,6 +92,11 @@ static struct memaccess_area nsec_shared[] = {
+#ifdef CFG_TEE_COHERENT_START
+register_phys_mem(MEM_AREA_TEE_COHERENT, CFG_TEE_COHERENT_START,
+					 CFG_TEE_COHERENT_SIZE);
+#endif
```
This should not be in commit "core: map registered coherent memory".
It should be in "core: linkable coherent memory section from CFG_TEE_COHERENT_START" which introduces `CFG_TEE_COHERENT_START`/`_SIZE`.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -842,49 +871,115 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+		init_smallpage_map(memory_map, vstart, coherent_start, false);
+		vstart = coherent_start + coherent_size;
+	}
+
```
Looks like the one and the one below could be merged...

etienne-lms commented:
Rebased (conflict to fix).
Tests in progress. qemu_virt ok, qemu_armv8 in progress... (in case ARMv8 would need coherent map support)

jenswi-linaro commented:
Please squash the commits as you'd like to have them merged and I'll take a final look.

etienne-lms commented:
`CFG_TEE_COHERENT_START`/`_SIZE` may not be a good label candidate.
- Any platform can get a coherent memory nicely mapped, using `register_phys_mem(MEM_AREA_TEE_COHERENT, ..., ...)`.
- Defining `CFG_TEE_COHERENT_START`/`_SIZE` makes the area a linked section inside the core implementation (allows static allocations inside coherent memory). 

Thus we may rather call this `CFG_LINKED_COHERENT_START`/`_SIZE`.

jenswi-linaro commented:
It think the define is good as it is.
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
@MrVan, this p-r proposes a generic mapping support for coherent RAM.
From abut to be merge https://github.com/OP-TEE/optee_os/pull/1729, your coherent memory does not need flat mapping support. My change may conflict with your use of coherent memory.

MrVan commented:
@etienne-lms I did not follow this p-r. I'll give a check and test with #1729.

etienne-lms commented:
Thanks

MrVan commented:
@etienne-lms 
It failed to boot correctly on i.mx7d-sdb
```
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: CFG_SHMEM_START type NSEC_SHM 0xbfe00000 size 0x00200000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: CFG_TA_RAM_START type TA_RAM 0xbe100000 size 0x01d00000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: VCORE_UNPG_RW_PA type TEE_RAM_RW 0xbe03d000 size 0x000c3000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: VCORE_UNPG_RX_PA type TEE_RAM_RX 0xbe000000 size 0x0003d000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: ROUNDDOWN(IRAM_S_BASE, CORE_MMU_DEVICE_SIZE) type TEE_COHERENT 0x00100000 size 0x00100000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: ROUNDDOWN(IRAM_BASE, CORE_MMU_DEVICE_SIZE) type TEE_COHERENT 0x00900000 size 0x00100000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: AIPS3_BASE type IO_SEC 0x30800000 size 0x00400000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: AIPS2_BASE type IO_SEC 0x30400000 size 0x00400000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: AIPS1_BASE type IO_SEC 0x30000000 size 0x00400000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: ANATOP_BASE type IO_SEC 0x30300000 size 0x00200000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:545: Physical mem map overlaps 0x30300000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: GIC_BASE type IO_SEC 0x31000000 size 0x00100000
DEBUG:   [0x0] TEE-CORE:add_phys_mem:532: CONSOLE_UART_BASE type IO_NSEC 0x30800000 size 0x00200000
DEBUG:   [0x0] TEE-CORE:verify_special_mem_areas:470: No NSEC DDR memory area defined
DEBUG:   [0x0] TEE-CORE:add_va_space:571: type RES_VASPACE size 0x00a00000
DEBUG:   [0x0] TEE-CORE:add_va_space:571: type SHM_VASPACE size 0x02000000
ERROR:   [0x0] TEE-CORE: assertion '!coherent_start' failed at core/arch/arm/mm/core_mmu.c:858 <init_mem_map>
ERROR:   [0x0] TEE-CORE: Panic at core/kernel/assert.c:50 <_assert_break>
ERROR:   [0x0] TEE-CORE: Call stack:
ERROR:   [0x0] TEE-CORE:  0xbe004ec5
ERROR:   [0x0] TEE-CORE:  0xbe00a907
ERROR:   [0x0] TEE-CORE:  0xbe009eab
ERROR:   [0x0] TEE-CORE:  0xbe0055d1
ERROR:   [0x0] TEE-CORE:  0xbe0000c4
```

jenswi-linaro commented:
+1 for `FLATMAP_COHERENT`

etienne-lms commented:
Closing deprecate p-r.
